ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs

```diff
@@ -0,0 +1,320 @@
+
+byronUtxosAreInsertsInShelleyUtxoDiff
+  :: LedgerState ByronBlock EmptyMK
+  -> LedgerState (ShelleyBlock Proto (ShelleyEra Crypto)) DiffMK
+  -> Bool
+byronUtxosAreInsertsInShelleyUtxoDiff srcLedgerState destLedgerState =
+    toNextUtxoDiff srcLedgerState == extractUtxoDiff destLedgerState
+  where
+    toNextUtxoDiff
+      :: LedgerState ByronBlock mk
+      -> Diff (TxIn Crypto) (Core.TxOut (ShelleyEra Crypto))
+    toNextUtxoDiff ledgerState =
+      let
+        Byron.UTxO utxo = Byron.cvsUtxo $ byronLedgerState ledgerState
+        keyFn = translateTxInByronToShelley . Byron.fromCompactTxIn
+        valFn = Diff.singletonInsert . translateCompactTxOutByronToShelley
+      in
+        Diff $ dimap keyFn valFn utxo
+
+    translateTxInByronToShelley :: Byron.TxIn -> TxIn Crypto
+    translateTxInByronToShelley byronTxIn =
+      let
+        Byron.TxInUtxo txId txIx = byronTxIn
+        shelleyTxId' = translateTxIdByronToShelley txId
+      in
+        TxIn shelleyTxId' (TxIx $ fromIntegral txIx)
+
+shelleyAvvmAddressesAreDeletesInUtxoDiff
+  :: LedgerState (ShelleyBlock Proto (ShelleyEra Crypto)) EmptyMK
+  -> LedgerState (ShelleyBlock Proto (AllegraEra Crypto)) DiffMK
+  -> Bool
+shelleyAvvmAddressesAreDeletesInUtxoDiff srcLedgerState destLedgerState =
+    toNextUtxoDiff srcLedgerState == extractUtxoDiff destLedgerState
+  where
+    toNextUtxoDiff
+      :: LedgerState (ShelleyBlock Proto (ShelleyEra Crypto)) EmptyMK
+      -> Diff (TxIn Crypto) (Core.TxOut (AllegraEra Crypto))
+    toNextUtxoDiff = avvmAddressesToUtxoDiff . stashedAVVMAddresses . shelleyLedgerState
+    avvmAddressesToUtxoDiff (UTxO m) =
+      let fn txOut = Diff.singletonDelete (Core.translateEra' () txOut)
+      in Diff $ dimap id fn m
+
+utxoTablesAreEmpty
+  :: LedgerState (ShelleyBlock srcProto srcEra) EmptyMK
+  -> LedgerState (ShelleyBlock destProto destEra) DiffMK
+  -> Bool
+utxoTablesAreEmpty _ destLedgerState = Diff.null $ extractUtxoDiff destLedgerState
+
+{-------------------------------------------------------------------------------
+    Generating tests for all translation predicates
+-------------------------------------------------------------------------------}
+
+-- | Predicate on pairs of ledger states of subsequent eras that are converted to property tests.
+data TranslationPredicate src dest = TranslationPredicate {
+    tpTestName :: TestName
+  , tpRun      :: LedgerState src EmptyMK -> LedgerState dest DiffMK -> Bool
+}
+
+-- Intermediate data type
+-- REVIEW: Can we get rid of this?
+data PreProperty x y = PreProperty {
+    ppTestName :: TestName
+  , ppRun      :: TestSetup x y -> Bool
+}
+
+toTestTrees
+  :: (AllPairs TestSetup Arbitrary xs, AllPairs TestSetup Show xs)
+  => InPairs TranslationPredicate xs
+  -> InPairs (RequiringBoth WrapLedgerConfig TranslateLedgerState) xs
+  -> [TestTree]
+toTestTrees predicates translations = mkTestTrees $ zipInPairs predicates translations
+  where
+    fn
```

Comment:
I would avoid naming this `fn` because we are heavily into `generics-sop` in these hard fork tests and there is actually a function named like that in SOP: https://hackage.haskell.org/package/generics-sop-0.5.1.2/docs/Generics-SOP.html#v:fn